src/Url.php

```diff
@@ -0,0 +1,243 @@
+<?php
+
+namespace Graby\HttpClient\Plugin\ServerSideRequestForgeryProtection;
+
+use Graby\HttpClient\Plugin\ServerSideRequestForgeryProtection\Exception\InvalidURLException;
+use Graby\HttpClient\Plugin\ServerSideRequestForgeryProtection\Exception\InvalidURLException\InvalidDomainException;
+use Graby\HttpClient\Plugin\ServerSideRequestForgeryProtection\Exception\InvalidURLException\InvalidIPException;
+use Graby\HttpClient\Plugin\ServerSideRequestForgeryProtection\Exception\InvalidURLException\InvalidPortException;
+use Graby\HttpClient\Plugin\ServerSideRequestForgeryProtection\Exception\InvalidURLException\InvalidSchemeException;
+
+class Url
+{
+    /**
+     * Validates the whole URL.
+     *
+     * @param string $url
+     * @param Options $options
+     *
+     * @throws InvalidURLException
+     *
+     * @return array
+     */
+    public static function validateUrl($url, Options $options)
+    {
+        if ('' === trim($url)) {
+            throw new InvalidURLException('Provided URL "'.$url.'" cannot be empty');
+        }
+
+        //Split URL into parts first
+        $parts = parse_url($url);
+
+        if (empty($parts)) {
+            throw new InvalidURLException('Error parsing URL "'.$url.'"');
+        }
+
+        if (!array_key_exists('host', $parts)) {
+            throw new InvalidURLException('Provided URL "'.$url.'" doesn\'t contain a hostname');
+        }
+
+        //If credentials are passed in, but we don't want them, raise an exception
+        if (!$options->getSendCredentials() && (array_key_exists('user', $parts) || array_key_exists('pass', $parts))) {
+            throw new InvalidURLException('Credentials passed in but "sendCredentials" is set to false');
+        }
+
+        //First, validate the scheme
+        if (array_key_exists('scheme', $parts)) {
+            $parts['scheme'] = self::validateScheme($parts['scheme'], $options);
+        } else {
+            //Default to http
+            $parts['scheme'] = 'http';
+        }
+
+        //Validate the port
+        if (array_key_exists('port', $parts)) {
+            $parts['port'] = self::validatePort($parts['port'], $options);
```

Comment:
Maybe you can cast the port as an integer.

See https://scrutinizer-ci.com/g/j0k3r/httplug-ssrf-plugin/inspections/b9d2643e-5c6a-4ffe-9c5f-31dcf0b966b0/issues/files/src/Url.php?status=new&orderField=path&order=asc&honorSelectedPaths=0

Reply:
I prefered updated the phpdoc of validatePort, parse_url return an int but the configuration of blacklist and whitelist use a string